parlai/mturk/core/worker_manager.py

```diff
@@ -0,0 +1,344 @@
+                    # another turker, record this as bad behavoir
+                    self.handle_bad_disconnect(worker_id)
+
+    def handle_bad_disconnect(self, worker_id):
+        """Update the number of bad disconnects for the given worker, block
+        them if they've exceeded the disconnect limit
+        """
+        if not self.is_sandbox:
+            self.mturk_workers[worker_id].disconnects += 1
+            self.disconnects.append({'time': time.time(), 'id': worker_id})
+            if self.mturk_workers[worker_id].disconnects > MAX_DISCONNECTS:
+                if self.opt['hard_block']:
+                    text = (
+                        'This worker has repeatedly disconnected from these '
+                        'tasks, which require constant connection to complete '
+                        'properly as they involve interaction with other '
+                        'Turkers. They have been blocked after being warned '
+                        'and failing to adhere. This was done in order to '
+                        'ensure a better experience for other '
+                        'workers who don\'t disconnect.'
+                    )
+                    self.mturk_manager.block_worker(worker_id, text)
+                    shared_utils.print_and_log(
+                        logging.INFO,
+                        'Worker {} blocked - too many disconnects'.format(
+                            worker_id
+                        ),
+                        True
+                    )
+                elif self.opt['disconnect_qualification'] is not None:
+                    self.mturk_manager.soft_block_worker(
+                        worker_id, 'disconnect_qualification')
+                    shared_utils.print_and_log(
+                        logging.INFO,
+                        'Worker {} soft blocked - too many disconnects'.format(
+                            worker_id
+                        ),
+                        True
+                    )
+
+    def worker_alive(self, worker_id):
+        """Creates a new worker record if it doesn't exist, returns state"""
+        if worker_id not in self.mturk_workers:
+            self.mturk_workers[worker_id] = WorkerState(worker_id)
+        return self.mturk_workers[worker_id]
+
+    def assign_task_to_worker(self, hit_id, assign_id, worker_id):
+        self.assignment_to_worker_id[assign_id] = worker_id
+        agent = self._create_agent(hit_id, assign_id, worker_id)
+        self.hit_id_to_agent[hit_id] = agent
+        curr_worker_state = self.mturk_workers[worker_id]
+        curr_worker_state.add_agent(agent)
+
+    def get_complete_hits(self):
+        """Returns the list of all currently completed HITs"""
+        hit_ids = []
+        for hit_id, agent in self.hit_id_to_agent.items():
+            if agent.hit_is_complete:
+                hit_ids.append(hit_id)
+        return hit_ids
+
+    def get_agent_work_status(self, assignment_id):
```

Comment:
should this function return something by default? or is `None` handled? looks like if there's an error that's not the one you specified, there's no return branch

Reply:
None of the callsites for this function do anything other than an `==` compare to `ASSIGNMENT_DONE` or `ASSIGNMENT_NOT_DONE`, so it hasn't come up, but yeah I should probably have this be an explicit return.